src/google/protobuf/stubs/common.h

```diff
@@ -132,6 +132,8 @@ namespace internal {
 // Checks if the buffer contains structurally-valid UTF-8.  Implemented in
 // structurally_valid.cc.
 PROTOBUF_EXPORT bool IsStructurallyValidUTF8(const char* buf, int len);
+PROTOBUF_EXPORT bool IsStructurallyValidUTF8(const char* buf, int len,
+                                             int* codepoints_consumed);
```

Comment:
> Why doesn't UTF8SpnStructurallyValid work for this case?

That is the same as `IsStructurallyValidUTF8` except it returns the number of valid bytes instead of just true or false. What we need is something the same as `IsStructurallyValidUTF8` except it also returns the number _codepoints_ consumed (not bytes, since some codepoints could be encoded with multiple bytes).

The count of codepoints is needed (vs. bytes) because we want the tokenizer's `col_` to report positions based on actual characters (which could be encoded with multiple bytes), not individual bytes. So after validating string and comment contents are valid UTF8, we'll need this codepoint count to adjust `col_`. We will also use the count if the comment or string has _invalid_ UTF8, so that the error reported for the offending data can indicate the correct location, in the event that the comment or string included multi-byte codepoints prior to the offending byte.

I'm happy to just press pause on this stuff if you need. I opened the PR in the hopes of maybe moving things forward (even if slowly) towards making sure that input source files are correct UTF8.

Reply:
Yea it would be great if we could pause on this for ~1 week or so while I finish opensourcing these.  Then we can take a closer look at porting this into the new repo